docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,921 @@
+- Remove `FailureReason` and `FailureMessage` to get rid of the confusing concept of terminal failures
+- Transition to new, improved, K8s API conventions aligned conditions
+
+Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
+After golang types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.
+
+```golang
+type MachineSetStatus struct {
+
+    // The number of ready replicas for this MachineSet. A machine is considered ready when Machine's Ready condition is true.
+    // +optional
+    ReadyReplicas int32 `json:"readyReplicas"`
+
+    // The number of up-to-date replicas for this MachineSet. A machine is considered up-to-date when Machine's UpToDate condition is true.
+    // +optional
+    UpToDateReplicas int32 `json:"upToDateReplicas"`
+
+    // Represents the observations of a MachineSet's current state.
+    // +listType=map
+    // +listMapKey=type
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+    // Other fields...
+    // NOTE: `FailureReason`, `FailureMessage` fields won't be there anymore
+}
+```
+
+| v1beta1 (current)                | v1beta2 (tentative Q1 2025)                              | v1beta1 removal (tentative Q1 2026) |
+|----------------------------------|----------------------------------------------------------|-------------------------------------|
+|                                  | `BackCompatibilty` (new)                                 | (removed)                           |
+| `ReadyReplicas` (deprecated)     | `BackCompatibilty.ReadyReplicas` (renamed) (deprecated)  | (removed)                           |
+| `ExprimentalReadyReplicas` (new) | `ReadyReplicas` (renamed)                                | `ReadyReplicas`                     |
+| `FailureReason` (deprecated)     | `BackCompatibilty.FailureReason` (renamed) (deprecated)  | (removed)                           |
+| `FailureMessage` (deprecated)    | `BackCompatibilty.FailureMessage` (renamed) (deprecated) | (removed)                           |
+| `Conditions`                     | `BackCompatibilty.Conditions` (renamed) (deprecated)     | (removed)                           |
+| `ExprimentalConditions` (new)    | `Conditions` (renamed)                                   | `Conditions`                        |
+| `UpToDateReplicas` (new)         | `UpToDateReplicas`                                       | `UpToDateReplicas`                  |
+| other fields...                  | other fields...                                          | other fields...                     |
+
+Notes:
+- The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).
```

Comment:
We could just not add any backward compatibility here and just try to support conversions instead?

Reply:
I took this idea from https://kubernetes.io/docs/reference/using-api/deprecation-policy/#fields-of-rest-resources.

I'm not opposed to other names (`v1beta1`, `BackCompatibility.v1beta1`, `Deprecated` etc), happy to take suggestions.

But IMO the real value of using this pattern for this transition is that allows us to design the ideal state, without caring about down conversions, which in my opinion could be tricky to implement (and frankly speaking not sure if worth the effort vs keeping old fields around for some more time).

Reply:
I think down conversion from v1beta2 conditions to v1beta1 conditions is only possible if the semantic is basically identical, because we can't really gather additional information in a conversion webhook. 

So I think it's the better option to have both fields side-by-side. It also makes it possible for folks to gradually migrate because they can access both the old and new conditions with both the v1beta1 and v1beta2 API types.

Reply:
> Were other names considered for the struct, would deprecated be a suitable alternative? 

Regarding naming. I would suggest `deprecatedV1beta1.conditions`, ...

Similarly I would suggest `v1beta2Conditions` or `v1beta2.conditions` (also for other fields like `v1beta2.readyReplicas` instead of `ExperimentalConditions` / `ExperimentalReadyReplicas`

Reply:
I'm fine with any of the suggested names (my preference goes to `V1beta2.*` and `V1beta1.Deprecated.*` for consistency) but I will wait for a few more +1 before changing this

Reply:
Fine for me!

Reply:
Leaning towards

```yaml
    # phase1
        status:
          conditions: ...
          readyReplicas: ...
          v1beta2: # new fields under a struct, not using "experimental"
              readyReplicas: ...
              conditions: ...          
    
    # phase2
      status:
        deprecated: # << old fields preserved for down conversion and moved under a deprecated struct
          v1beta1:  # << additional level of nesting to make it extensible
              readyReplicas: ...
              conditions: ...
        readyReplicas: ...
        conditions: ...
    
    # phase3
      status:
        readyReplicas: ...
        conditions: ...
```